packages/smooth_app/lib/pages/product/new_product_page.dart

```diff
@@ -0,0 +1,274 @@
+  State<NewProductPage> createState() => _ProductPageState();
+}
+
+enum ProductPageMenuItem { WEB, REFRESH }
+
+class _ProductPageState extends State<NewProductPage> {
+  late Product _product;
+
+  @override
+  void initState() {
+    super.initState();
+    _updateHistory(context);
+    _product = widget.product;
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    final LocalDatabase localDatabase = context.watch<LocalDatabase>();
+    final AppLocalizations appLocalizations = AppLocalizations.of(context)!;
+    return Scaffold(
+      backgroundColor: SmoothTheme.COLOR_PRODUCT_PAGE_BACKGROUND,
+      appBar: AppBar(
+        title: Text(_getProductName(appLocalizations)),
+        actions: <Widget>[
+          PopupMenuButton<ProductPageMenuItem>(
+            itemBuilder: (BuildContext context) =>
+                <PopupMenuEntry<ProductPageMenuItem>>[
+              PopupMenuItem<ProductPageMenuItem>(
+                value: ProductPageMenuItem.WEB,
+                child: Text(appLocalizations.label_web),
+              ),
+              PopupMenuItem<ProductPageMenuItem>(
+                value: ProductPageMenuItem.REFRESH,
+                child: Text(appLocalizations.label_refresh),
+              ),
+            ],
+            onSelected: (final ProductPageMenuItem value) async {
+              switch (value) {
+                case ProductPageMenuItem.WEB:
+                  LaunchUrlHelper.launchURL(
+                      'https://openfoodfacts.org/product/${_product.barcode}/',
+                      false);
+                  break;
+                case ProductPageMenuItem.REFRESH:
+                  _refreshProduct(localDatabase, appLocalizations);
+                  break;
+                default:
+                  throw UnimplementedError(
+                      "$value Popup item isn't handled yet.");
+              }
+            },
+          ),
+        ],
+      ),
+      body: _buildProductBody(context),
+    );
+  }
+
+  Future<void> _refreshProduct(
+      LocalDatabase localDatabase, AppLocalizations appLocalizations) async {
+    final ProductDialogHelper productDialogHelper = ProductDialogHelper(
+      barcode: _product.barcode!,
+      context: context,
+      localDatabase: localDatabase,
+      refresh: true,
+    );
+    final Product? product =
+        await productDialogHelper.openUniqueProductSearch();
+    if (product == null) {
+      productDialogHelper.openProductNotFoundDialog();
+      return;
+    }
+    ScaffoldMessenger.of(context).showSnackBar(
+      SnackBar(
+        content: Text(appLocalizations.product_refreshed),
+        duration: const Duration(seconds: 2),
+      ),
+    );
+    setState(() {
+      _product = product;
+    });
+    await _updateHistory(context);
+  }
+
+  Future<void> _updateHistory(BuildContext context) async {
+    final LocalDatabase localDatabase = context.read<LocalDatabase>();
+    await DaoProductExtra(localDatabase).putLastSeen(widget.product);
```

Comment:
copy pasted code but I think you've caught a bug here, should definitely use `_product` instead of `widget.product`. Fortunately, the bug is benign and harmless which is why the app didn't complain because of it.

Both `_product` and `_updateHistory` (which I now think was poorly named) do get updated together, but I can't manage to put them together in a method, conceptually it doesn't make sense, what would the method be called ? (_updateProductAndLastSeenProductHistory?) generally I'd like if methods only do 1 unit of work.

Anyways, great catch, I decided to pass in product as a parameter, no point using a variable that is getting updated at the same time and making assumption about it's state.

Reply:
How about `_setProduct()`?

Reply:
Sorry I missed this comment and already merged the change, `_setProduct` would then also be updating the local database with product history which I think becomes a side effect of the function, which isn't great IMO.

Reply:
Ack.

After thinking about this some more, I think a better way to structure this could be to have a separate parent component managing the retrieval and refresh of `Product`. That component would call `NewProductPage` like so:
```dart
NewProductPage(product, onRefresh: () { ... })
```